packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp

```diff
@@ -70,17 +72,147 @@ class Distributor;
+/// \tparam LocalMatrixType Type of the local matrix
+///
+template<class GO, class ViewType, class ExportLIDsViewType, class LocalMatrixType>
+struct NumPacketsFunctor {
+
+  typedef typename ViewType::value_type value_type;
+  typedef typename LocalMatrixType::row_map_type row_map_type;
+  typedef typename LocalMatrixType::value_type IST;
+  typedef typename LocalMatrixType::ordinal_type LO;
+
+  ViewType num_packets;
+  ViewType num_ent_per_row;
+  ExportLIDsViewType export_lids;
+  row_map_type row_map;
+
+  NumPacketsFunctor(ViewType num_packets, ViewType num_ent_per_row,
+      ExportLIDsViewType export_lids, LocalMatrixType local_matrix):
+    num_packets(num_packets), num_ent_per_row(num_ent_per_row),
+    export_lids(export_lids), row_map(local_matrix.graph.row_map) {};
+
+  KOKKOS_INLINE_FUNCTION
+    void operator() (size_t i) const {
+      size_t io = static_cast<size_t>(export_lids[i]);
+      const value_type count = static_cast<value_type>(row_map[io+1] - row_map[io]);
+      const value_type num_bytes = sizeof(LO) + count * (sizeof(IST) + sizeof(GO));
+      num_ent_per_row[i] = count;
+      num_packets[i] = (count == 0) ? 0 : num_bytes;
+    }
+};
+
+template<class ViewType, class ExportLIDsViewType,
+         class LocalMatrixType, class LocalMapType>
+struct PackCrsMatrixFunctor {
+
+  typedef typename LocalMapType::local_ordinal_type LO;
+  typedef typename LocalMapType::global_ordinal_type GO;
+  typedef typename LocalMatrixType::value_type IST;
+
+  static_assert (std::is_same<LO, typename LocalMatrixType::ordinal_type>::value,
+                 "LocalMapType::local_ordinal_type and "
+                 "LocalMatrixType::ordinal_type must be the same.");
+
+  ViewType num_packets_per_lid;
+  ViewType num_ent_per_row;
+  ViewType offsets;
+  char* const exports_raw_ptr;
+  ExportLIDsViewType export_lids;
+  const size_t buf_size;
+  LocalMatrixType local_matrix;
+  LocalMapType local_col_map;
+
+  size_t first_bad_index; // only valid if outOfBounds == true.
+  size_t first_bad_offset;   // only valid if outOfBounds == true.
+  size_t first_bad_num_bytes; // only valid if outOfBounds == true.
+  bool out_of_bounds;
+  //bool pack_error;
+
+  PackCrsMatrixFunctor(ViewType num_packets_per_lid, ViewType num_ent_per_row,
+      ViewType offsets, char* const exports_raw_ptr, ExportLIDsViewType export_lids,
+      const size_t buf_size, LocalMatrixType local_matrix, LocalMapType local_col_map):
+    num_packets_per_lid(num_packets_per_lid), num_ent_per_row(num_ent_per_row),
+    offsets(offsets), exports_raw_ptr(exports_raw_ptr), export_lids(export_lids),
+    buf_size(buf_size), local_matrix(local_matrix), local_col_map(local_col_map)
+  {
+    first_bad_index = 0;
+    first_bad_offset = 0;
+    first_bad_num_bytes = 0;
+    out_of_bounds = false;
+   // pack_error = false;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+    void operator()(size_t i, int& error_code) const {
```

Comment:
Don't use size_t for the loop index here, because you know that the number of rows to pack must fit in LO.